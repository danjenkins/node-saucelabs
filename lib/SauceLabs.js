var crypto      = require('crypto');
var https       = require('https');
var querystring = require('querystring');
var utils       = require('./utils');
var HttpsProxyAgent = require('https-proxy-agent');
var url = require('url');



var extend  = utils.extend;
var replace = utils.replace;

var DEFAULTS = {
  username:  null,
  password:  null,
  proxy: null,
  hostname:  'saucelabs.com',
  base:      '/rest/v1/',
  port:      '443',
  requestLog: null
};

function SauceLabs(options, headers) {
  this.options = extend({}, DEFAULTS, options);
  this.options.auth = this.options.username + ':' + this.options.password;
  this.headers = extend({}, headers);
}

module.exports = SauceLabs;

// API

SauceLabs.prototype.getAccountDetails = function (callback) {
  this.send({
    method: 'GET',
    path: 'users/:username'
  }, callback);
};

SauceLabs.prototype.getAccountLimits = function (callback) {
  this.send({
    method: 'GET',
    path: ':username/limits'
  }, callback);
};

SauceLabs.prototype.getUserActivity = function (start, end, callback) {
  if (typeof start === 'function') {
    callback = start;
    start    = null;
    end      = null;
  } else if (typeof end === 'function') {
    callback = end;
    end      = null;
  }

  var dates = (start != null || end != null) ? {} : null;
  if (start != null) {
    dates.start = formatDate(start);
  }
  if (end != null) {
    dates.end = formatDate(end);
  }

  this.send({
    method: 'GET',
    path: ':username/activity',
    query: dates
  }, callback);
};

SauceLabs.prototype.getAccountUsage = function (callback) {
  this.send({
    method: 'GET',
    path: 'users/:username/usage'
  }, callback);
};

SauceLabs.prototype.getJobs = function (callback) {
  this.send({
    method: 'GET',
    path: ':username/jobs',
    query: { full: true }
  }, callback);
};

SauceLabs.prototype.showJob = function (id, callback) {
  this.send({
    method: 'GET',
    path: ':username/jobs/:id',
    args: { id: id }
  }, callback);
};

SauceLabs.prototype.showJobAssets = function (id, callback) {
  this.send({
    method: 'GET',
    path: ':username/jobs/:id/assets',
    args: { id: id }
  }, callback);
};

SauceLabs.prototype.updateJob = function (id, data, callback) {
  this.send({
    method: 'PUT',
    path: ':username/jobs/:id',
    args: { id: id },
    data: data
  }, callback);
};

SauceLabs.prototype.stopJob = function (id, data, callback) {
  this.send({
    method: 'PUT',
    path: ':username/jobs/:id/stop',
    args: { id: id },
    data: data
  }, callback);
};

SauceLabs.prototype.getActiveTunnels = function (callback) {
  this.send({
    method: 'GET',
    path: ':username/tunnels'
  }, callback);
};

SauceLabs.prototype.getTunnel = function (id, callback){
  this.send({
    method: 'GET',
    path: ':username/tunnels/:id',
    args: { id: id }
  }, callback);
};

SauceLabs.prototype.deleteTunnel = function (id, callback){
  this.send({
    method: 'DELETE',
    path: ':username/tunnels/:id',
    args: { id: id }
  }, callback);
};

SauceLabs.prototype.getServiceStatus = function (callback) {
  this.send({
    method: 'GET',
    path: 'info/status'
  }, callback);
};

SauceLabs.prototype.getBrowsers = function (callback) {
  this.send({
    method: 'GET',
    path: 'info/browsers'
  }, callback);
};

SauceLabs.prototype.getAllBrowsers = function (callback) {
  this.send({
    method: 'GET',
    path: 'info/browsers/all'
  }, callback);
};

SauceLabs.prototype.getSeleniumBrowsers = function (callback) {
  this.send({
    method: 'GET',
    path: 'info/browsers/selenium-rc'
  }, callback);
};

SauceLabs.prototype.getWebDriverBrowsers = function (callback) {
  this.send({
    method: 'GET',
    path: 'info/browsers/webdriver'
  }, callback);
};

SauceLabs.prototype.getTestCounter = function (callback) {
  this.send({
    method: 'GET',
    path: 'info/counter'
  }, callback);
};

SauceLabs.prototype.updateSubAccount = function (data, callback) {
  this.send({
    method: 'POST',
    path: 'users/:username/subscription',
    data: data
  }, callback);
};

SauceLabs.prototype.deleteSubAccount = function (callback) {
  this.send({
    method: 'DELETE',
    path: 'users/:username/subscription'
  }, callback);
};

SauceLabs.prototype.createSubAccount = function (data, callback) {
  this.send({
    method: 'POST',
    path: 'users/:username',
    data: data
  }, callback);
};

SauceLabs.prototype.createPublicLink = function (id, date, useHour, callback) {
  if (typeof date === 'function') {
    callback = date;
    date     = null;
    useHour  = false;
  } else if (typeof useHour === 'function') {
    callback = useHour;
    useHour  = false;
  }

  if (date != null) {
    date = formatDate(date, useHour);
  }

  var link = generateLink(this.options.hostname, this.options.auth, date, id);
  callback(null, link);
};

SauceLabs.prototype.uploadFile = function (filename, data, overwrite, callback) {
  this.send({
      method: 'POST',
      path: 'storage/:username/:filename',
      args: {filename: filename},
      query: {overwrite: overwrite},
      data: data
    }, callback);
};

SauceLabs.prototype.getStoredFiles = function (callback) {
  this.send({
    method: 'GET',
    path: 'storage/:username'
  }, callback);
};

SauceLabs.prototype.send = function (message, callback) {
  var method  = message.method,
      path    = message.path,
      args    = message.args,
      query   = message.query,
      data    = message.data,
      body    = Buffer.isBuffer(data) ? data: JSON.stringify(data);

  // Build path with base, placeholders, and query.
  path = this.options.base + replace(path, extend({}, this.options, args));
  if (query != null) {
    path += '?' + querystring.stringify(query);
  }

  // Make the request.
  var options = extend({}, this.options, {
    method: method,
    path:   path,

    headers: extend({
      'Accept':         'application/json',
      'Content-Type':   Buffer.isBuffer(data) ? 'application-octetstream' : 'application/json',
      'Content-Length': body != null ? body.length : 0
    }, this.headers)
  });
  makeRequest(options, body, callback);
};

// Helpers

function formatDate(date, useHour) {
  return date.toISOString().replace(/T(\d+).*/, (useHour ? '-$1' : ''));
}

function generateToken(auth, date, job) {
  var key = auth + (date ? ':' + date : '');
  return crypto
    .createHmac('md5', key)
    .update(job)
    .digest('hex');
}

function generateLink(hostname, auth, date, job) {
  return replace('https://:hostname/jobs/:id?auth=:token', {
    hostname: hostname,
    id:       job,
    token:    generateToken(auth, date, job)
  });
}

function makeRequest(options, body, callback) {
    if(options.proxy){
        options.agent = new HttpsProxyAgent(url.parse(options.proxy));
    }
    var request = https.request(options, function (response) {
    var result = '';
    response
      .on('data', function (chunk) {
        result += chunk;
      })
      .on('end', function () {
        var res;

        try {
          res = JSON.parse(result);
        } catch (e) {

          var err = 'Could not parse response: ' + result;
          if(typeof options.requestLog === 'function') {
            options.requestLog(err, request);
          }
          callback(err);
          return;
        }

        if(typeof options.requestLog === 'function') {
          options.requestLog(null, request);
        }

        if (response.statusCode === 200) {
          callback(null, res);
        } else {
          callback(res);
        }
      });
  });

  request
    .on('error', function (err) {
      callback('Could not send request: ' + err.message);
    });

  if (body != null) {
    request.write(body);
  }

  request.end();
}
