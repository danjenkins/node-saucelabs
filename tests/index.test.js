import util from 'util'
import got from 'got'

import SauceLabs from '../src'

jest.mock('fs')
const fs = require('fs')

test('should be inspectable', () => {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    expect(util.inspect(api)).toContain(`{
  username: 'foo',
  key: 'XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXbar',
  region: 'us',
  headless: false,
  proxy: undefined
}`)
})

test('should have to string tag', () => {
    expect(Object.prototype.toString.call(
        new SauceLabs({ user: 'foo', key: 'bar' })
    )).toBe('[object SauceLabs API Client]')
})

test('should not provide an iterator', () => {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    expect(() => [...api]).toThrow('is not iterable')
})

test('should return public properties', () => {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    expect(api.region).toBe('us')
    expect(api.username).toBe('foo')
    expect(api._accessKey).toBe(undefined)
})

test('should grab username and access key from env variable', () => {
    jest.resetModules()
    process.env.SAUCE_USERNAME = 'barfoo'
    process.env.SAUCE_ACCESS_KEY = 'foobar'
    const SauceLabsNew = require('../src').default
    const api = new SauceLabsNew()
    expect(util.inspect(api))
        .toContain('username: \'barfoo\'')
    expect(util.inspect(api))
        .toContain('key: \'XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXfoobar\'')
})

test('should grab proxy from env variable', () => {
    jest.resetModules()
    process.env.HTTP_PROXY = 'http://my.proxy.com:8080'
    const SauceLabsNew = require('../src').default
    const api = new SauceLabsNew()
    expect(util.inspect(api))
        .toContain('proxy: \'http://my.proxy.com:8080\'')
})

test('should throw if API command is unknown', () => {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    expect(() => api.doSomethingCool(123, { foo: 'bar' }))
        .toThrow('Couldn\'t find API endpoint for command "doSomethingCool"')
})

test('should allow to call an API method with param in url', async () => {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    await api.getUserConcurrency('someuser')
    expect(got.mock.calls[0][0])
        .toBe('https://saucelabs.com/rest/v1.1/users/someuser/concurrency')
})

test('should allow to call an API method with param as option', async () => {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    await api.listJobs('someuser', {
        limit: 123,
        full: true
    })

    const uri = got.mock.calls[0][0]
    const req = got.mock.calls[0][1]
    expect(uri).toBe('https://saucelabs.com/rest/v1.1/someuser/jobs')
    expect(req.searchParams).toEqual({
        limit: 123,
        full: true
    })
})

test('should allow to make a request with body param', async () =>  {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    await api.updateJob('foobaruser', '690c5877710c422d8be4c622b40c747f', { passed: true })

    const req = got.mock.calls[0][1]
    expect(got.put).toBeCalled()
    expect(req.json).toEqual({ passed: true })
})

test('should allow to make a request with body param via CLI call', async () =>  {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    await api.updateJob('foobaruser', '690c5877710c422d8be4c622b40c747f', '{ "passed": false }')

    const req = got.mock.calls[0][1]
    expect(got.put).toBeCalled()
    expect(req.json).toEqual({ passed: false })
})

test('should update RDC job status with body param', async () =>  {
    const api = new SauceLabs({})
    await api.updateTest('89ec3cca-7092-41f1-8037-d035579fb8d1', { passed: true })

    const req = got.mock.calls[0][1]
    expect(got.put).toBeCalled()
    expect(req.json).toEqual({ passed: true })
})

test('should update RDC job status with body param via CLI call', async () =>  {
    const api = new SauceLabs({})
    await api.updateTest('89ec3cca-7092-41f1-8037-d035579fb8d1', '{ "passed": false }')

    const req = got.mock.calls[0][1]
    expect(got.put).toBeCalled()
    expect(req.json).toEqual({ passed: false })
})

test('should fail if param has wrong type', async () => {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    const error = await api.listJobs(123, {
        limit: 123,
        full: true
    }).catch((err) => err)
    expect(error.message).toBe('Expected parameter for url param \'username\' from type \'string\', found \'number\'')
})

test('should fail if option has wrong type', async () => {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    const error = await api.listJobs('someuser', {
        limit: '123',
        full: true
    }).catch((err) => err)
    expect(error.message).toBe('Expected parameter for option \'limit\' from type \'number\', found \'string\'')
})

test('should handle error case', async () => {
    const response = new Error('Not Found')
    response.statusCode = 404
    response.body = { message: 'Not Found' }
    got.get.mockReturnValueOnce(Promise.reject(response))
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    const error = await api.listJobs('someuser', {
        limit: 123,
        full: true
    }).catch((err) => err)
    expect(error.message).toBe('Failed calling listJobs: Not Found')
})

test('should be able to download assets', async () => {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    await api.downloadJobAsset('some-id', 'performance.json')
    const uri = got.get.mock.calls[0][0]
    expect(uri).toContain('https://assets.saucelabs.com/jobs/some-id/performance.json')
    expect(uri).toContain('auth=a2600100e3d1990721be97c093f64567')
})

test('should handle errors when downloading assets', async () => {
    const response = new Error('Not Found')
    response.statusCode = 404
    response.body = { message: 'Not Found' }
    got.get.mockReturnValueOnce(Promise.reject(response))

    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    const error = await api.downloadJobAsset('some-id', 'performance.json').catch((err) => err)
    expect(error.message).toBe('There was an error downloading asset performance.json: Not Found')
})

test('should parse text responses if headers expect json', async () => {
    const reqRespond = { foo: 'bar' }
    got.get.mockReturnValueOnce(Promise.resolve({
        headers: {
            'content-type': 'application/json'
        },
        body: JSON.stringify(reqRespond)
    }))
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    const result = await api.downloadJobAsset('some-id', 'performance.json')
    expect(result).toEqual(reqRespond)
})

test('should fail if parameters are not given properly', async () => {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    const error = new Error('You need to define a job id and the file name of the asset as a string')
    await expect(api.downloadJobAsset()).rejects.toEqual(error)
    await expect(api.downloadJobAsset('foo')).rejects.toEqual(error)
    await expect(api.downloadJobAsset(123, 'bar')).rejects.toEqual(error)
})

test('should support proxy options', async () => {
    const proxy = 'http://my.proxy.com:8080'
    const api = new SauceLabs({ user: 'foo', key: 'bar', proxy })
    await api.downloadJobAsset('some-id', 'performance.json')
    const requestOptions = got.extend.mock.calls[1][0]

    await expect(requestOptions.agent).toBeDefined()
})

test('should put asset into file as binary', async () => {
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    await api.downloadJobAsset('some-id', 'performance.json', { filepath: '/asset.json' })
    expect(fs.writeFileSync).toBeCalledWith('/asset.json', undefined, { encoding: 'binary' })
})

test('should put asset into file as json file', async () => {
    got.setHeader({ 'content-type': 'application/json' })
    const api = new SauceLabs({ user: 'foo', key: 'bar' })
    await api.downloadJobAsset('some-id', 'performance.json', { filepath: '/asset.json' })
    expect(fs.writeFileSync).toBeCalledWith('/asset.json', undefined, { encoding: 'utf8' })
})

afterEach(() => {
    fs.writeFileSync.mockClear()
    got.mockClear()
    got.extend.mockClear()
    got.put.mockClear()
    got.get.mockClear()
})
