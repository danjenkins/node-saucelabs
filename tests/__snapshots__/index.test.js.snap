// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`startSauceConnect should start sauce connect with proper parsed args 1`] = `
Array [
  Array [
    "/foo/bar",
    Array [
      "--proxy-tunnel=abc",
      "--se-port=1234",
      "--user=foo",
      "--api-key=bar",
      "--rest-url=https://us-east-1.saucelabs.com/rest/v1",
    ],
  ],
]
`;
